### DISCUSSION TITLE
Partitions and idle mode

### DISCUSSION THREAD

**Discussion by jcorporation** [MusicPlayerDaemon/MPD#1521] [Q&A]
I played around with partitions, especially with the idle feature in combination with partitions.

It seems MPD emits all idle events to all partitions and I found no approach to determine from which partition the idle event is triggered.

I think idle events should reference the triggering partition.

**Comment by MiM-TMassey** [comment]
I can’t say much about the protocol issues. I just wanted to say that useful partition support would be *very* useful.  I’m using MPD with independent speakers in about a dozen zones in and around my house.  With outputs, I can turn individual rooms on and off, which is great. But with partitions, now different people can listen to different things and turn them on and off throughout the house.  Even my non-tech family members are excited about that!  Being able to do that from a single client would make managing that so much easier.  But finding clients with partition support is difficult.  MyMPD is a great client already: I hope that MPD and MyMPD can work together to come up with a partition mechanism that works for everyone.

**Reply by sevmonster** [reply]
Your idea is pretty sick. For me, I would like to be able to stream to multiple Icecast/PulseAudio endpoints that are independently controlled, but the end-user experience is quite poor and involved.

**Comment by jcorporation** [comment]
I found commit 49309b419f47d7acd4ba3ab2c6d1923df875587f:

>  Partition: add a local idle_monitor
Make idle events per-partition, but leave Instance::EmitIdle() and its
underlying idle_monitor which broadcasts idle events to all
partitions.

Idle events are already partition specific but broadcasted to all partitions? Is this only for backward compatibility? It would be great to keep partition specific events inside the partition. 

@MaxKellermann: Could this be implemented? E.g. with a config option to change the behavior.

**Reply by MaxKellermann** [reply]
I don't know how to answer to your post, because what you want is implemented, and you didn't post any information about which idle event was broadcasted to all partitions that should not have been. But if you see that, write a bug report. There may be bugs, or some code isn't yet separated out yet, and in any case, that will be fixed.

**Reply by jcorporation** [reply]
Thanks for your reply. I will collect the logs and open a new issue.

I tested the partition idle event behavior with the player idle event. I suppose that this event is partition specific?

**Reply by MaxKellermann** [reply]
Yes, every partition has its own player. But there are indeed lots of callers which invoke the global `idle_add()` function instead of `Partition::EmitIdle()`. That's just an incomplete transition; no client should ever be confused or buggy because it receives too many idle events; at worst, this is some unnecessary overhead. This is worth optimizing, but not a user-visible problem.

**Reply by MaxKellermann** [reply]
Commit 64f84d5468d97b6de7fed2c36b56b57d3a82b6cd reduces the number of global idle events. Some more work needs to be done. `git grep idle_add` shows candidates.

**Reply by jcorporation** [reply]
Great, thanks! I will have a look at it.

**Reply by jcorporation** [reply]
Looked through `git grep idle_add` output and prepared some pull requests.

Pull request #1608:
IDLE_MIXER events from `src/output/OutputCommand.cxx`

Pull request #1604:
IDLE_PLAYER and IDLE_OPTIONS events from `src/player/Control.cxx`

There are some global IDLE_OUTPUT events remaining in `src/output/OutputCommand.cxx`, but I could not figure out how to replace these with partition specific events. I am not so familiar with C++ and the MPD codebase.
